### DISCUSSION TITLE
Using Colima & Docker

### DISCUSSION THREAD

**Discussion by eratio08** [mintoolkit/mint#45] [Q&A] [answered]
Hello I was trying to use this on macOS using Colima.
I'm getting the following error

```bash
﻿level=error msg="dockerutil.BuildEmptyImage: dockerapi.BuildImage() error = API error (400): client version 1.12 is too old. Minimum supported API
 version is 1.24, please upgrade your client to a newer version / output: ...
```

This is the output of `mint version`
```bash
﻿❯ mint version
cmd=version info=app location='/usr/local/bin' status='OUTDATED' current='1.41.4' verdict='your installed version is OUTDATED (local=1.41.5 current=1.41.4)' cmd='version' version=
'darwin/arm64|Aurora|1.41.5|4cc2b185c9cdcd6d0586246e21d8aecc5d847feb|2024-07-01_04:36:47AM' container='false' dsimage='false' 
cmd=version info=host cmd='version' osname='Sonoma (14.6)' osbuild='23G80' version=' Fri Jul  5 17:53:24 PDT 2024; root:xnu-10063.141.1~2/RELEASE_ARM64_T6020' release='23.6.0' sys
name='darwin' 
cmd=version info=docker cmd='version' name='colima' kernel.version='6.8.0-36-generic' operating.system='Ubuntu 24.04 LTS' ostype='linux' server.version='27.1.1' architecture='aarc
h64' 
cmd=version info=dclient cmd='version' api.version='1.46' min.api.version='1.24' build.time='2024-07-23T20:00:07.000000000+00:00' git.commit='cc13f95' 
```
(unrelated: the reported latest version is wrong for some reason it's `1.41.5` not `1.41.4`)

and `docker version`

```bash
Client: Docker Engine - Community
 Version:           27.1.1
 API version:       1.46
 Go version:        go1.22.5
 Git commit:        63125853e3
 Built:             Fri Jul 19 17:35:01 2024
 OS/Arch:           darwin/arm64
 Context:           colima

Server: Docker Engine - Community
 Engine:
  Version:          27.1.1
  API version:      1.46 (minimum version 1.24)
  Go version:       go1.21.12
  Git commit:       cc13f95
  Built:            Tue Jul 23 20:00:07 2024
  OS/Arch:          linux/arm64
  Experimental:     false
 containerd:
  Version:          1.7.19
  GitCommit:        2bf793ef6dc9a18e00cb12efb64355c2c9d5eb41
 runc:
  Version:          1.7.19
  GitCommit:        v1.1.13-0-g58aa920
 docker-init:
  Version:          0.19.0
  GitCommit:        de40ad0
```

The error message hint that the client (mint) uses an api version `1.12` which is not supported by my local docker engine.

Is there any way I can work around this/fix it?

**Comment by kcq** [comment, marked as answer]
@eratio08 need to debug this behavior... because the (Docker) API version should default to `1.25`. You should be able to use the global `--crt-api-version` flag to select an API version to use. There's also a native way to do it with the `DOCKER_API_VERSION` environment variable (try it if `--crt-api-version` doesn't work for you)

**Reply by eratio08** [reply]
Hey @kcq thanks for the repose! I will tests these and report if any of these worked! 🙌

**Reply by eratio08** [reply]
Using 
```bash
﻿mint slim --target some-image --registry-account AWS --registry-secret $DOCKER_PW --crt-api-version=1.46

Incorrect Usage: flag provided but not defined: -crt-api-version
```

does not work properly. I also tried supplying it using `--crt-api-version 1.46` but got the same error.

Using `DOCKER_API_VERSION` worked ✅ Thanks!

**Reply by kcq** [reply]
--crt-api-version is a global flag and it needs to be provided before the command name ("slim"). Either way, it's good that the environment variable worked.

**Comment by eratio08** [comment]
@kcq I've also noticed that `mint` does not recognizes/utilizes docker contexts and only looks for the docker socket expose via the env var `DOCKER_HOST`. The context are used by Colima to provide the alternative socket path.

**Reply by dan-jacobson** [reply]
I've noticed the same thing using `Orbstack`. It would be nice for `mint` to respect `DOCKER_CONTEXT`, or even just try some sane defaults (I have my active context symlinked to `/var/run/docker.sock` but for some reason `mint` didn't pick that up??

**Reply by kcq** [reply]
@eratio08 / @dan-jacobson for now you need to explicitly set the socket address with the global `--host` flag. For example, this is what I did with my Orbstack setup: `mint --host unix:///Users/q/.orbstack/run/docker.sock slim nginx`

Hopefully contexts will make it in the next release. Keeping my fingers crossed :-)

**Reply by dan-jacobson** [reply]
Interesting -- that doesn't work for me.

Running `docker context ls` I get:
```
NAME         DESCRIPTION                               DOCKER ENDPOINT                               ERROR
default                                                           unix:///var/run/docker.sock                   
orbstack *   OrbStack                                  unix:///Users/drj/.orbstack/run/docker.sock   
```
but then running `mint --host unix:///Users/drj/.orbstack/run/docker.sock slim nginx` I get:
```
cmd=slim info=param.http.probe message='using default probe' 
cmd=slim error=docker.connect.error message='missing Docker connection info'
cmd=slim state=exited code=16777219 version=darwin/arm64|Aurora|1.41.6|53e8072bb239bc3cad5b26a14f3ab8b678db0c9a|2024-08-25_03:34:13AM location=/usr/local/bin 
cmd=slim info=report file='slim.report.json' 
cmd=slim info=exit code='16777219' version='darwin/arm64|Aurora|1.41.6|53e8072bb239bc3cad5b26a14f3ab8b678db0c9a|2024-08-25_03:34:13AM' location='/usr/local/bin'
```

**Reply by kcq** [reply]
@dan-jacobson can you help out with debugging this? try enabling trace logging: `mint --log-level=trace --host unix:///Users/drj/.orbstack/run/docker.sock slim nginx` This should print a lot of stuff including the docker client creation trace, which should provide clues why this happens.

**Reply by dan-jacobson** [reply]
Appreciate the help :) -- I ran `mint --log-level=trace --host unix:///Users/drj/.orbstack/run/docker.sock slim nginx` and here's what I see:

```
time="2024-09-26T09:31:02-04:00" level=debug msg="sysinfo => system.SystemInfo{Sysname:\"darwin\", Nodename:\"Daniels-MacBook-Pro-3.local\", Release:\"23.6.0\", Version:\" Mon Jul 29 21:14:30 PDT 2024; root:xnu-10063.141.2~1/RELEASE_ARM64_T6000\", Machine:\"arm64\", Domainname:\"\", OsBuild:\"23G93\", Distro:system.DistroInfo{Name:\"\", Version:\"\", DisplayName:\"Sonoma (14.6)\"}}"
time="2024-09-26T09:31:02-04:00" level=debug msg="param.error.app.options - no app params"
cmd=slim info=param.http.probe message='using default probe' 
time="2024-09-26T09:31:02-04:00" level=debug msg="customImageTag='', additionalTags=[]string(nil)" app=mint cmd=slim
cmd=slim error=docker.connect.error message='missing Docker connection info'
cmd=slim state=exited code=16777219 version=darwin/arm64|Aurora|1.41.6|53e8072bb239bc3cad5b26a14f3ab8b678db0c9a|2024-08-25_03:34:13AM location=/usr/local/bin 
cmd=slim info=report file='slim.report.json' 
cmd=slim info=exit version='darwin/arm64|Aurora|1.41.6|53e8072bb239bc3cad5b26a14f3ab8b678db0c9a|2024-08-25_03:34:13AM' location='/usr/local/bin' code='16777219'
```

I'm not really seeing any clues in there. What do you see?

The `slim.report.json` doesn't give me any guidance either:

```
{
 "document": "doc.report.command",
 "version": "ov/command/slim/1.1",
 "engine": "darwin/arm64|Aurora|1.41.6|53e8072bb239bc3cad5b26a14f3ab8b678db    0c9a|2024-08-25_03:34:13AM",
 "containerized": false,
 "host_distro": {
    "name": "",
    "version": "",
    "display_name": "Sonoma (14.6)"
  },
  "type": "slim",
  "state": "error",
  "error": "docker.connect.error",
  "target_reference": "nginx",
  "system": {
    "type": "",
    "release": "",
    "distro": {
      "name": "",
      "version": "",
      "display_name": ""
    }
  },
  "source_image": {
    "identity": {
      "id": ""
    },
    "size": 0,
    "size_human": "",
    "create_time": "",
    "architecture": "",
    "container_entry": {
      "exe_path": ""
    }
  },
  "minified_image_size": 0,
  "minified_image_size_human": "",
  "minified_image": "",
  "minified_image_id": "",
  "minified_image_digest": "",
  "minified_image_has_data": false,
  "minified_by": 0,
  "artifact_location": "",
  "container_report_name": "",
  "seccomp_profile_name": "",
  "apparmor_profile_name": "",
  "image_stack": null,
  "image_created": false,
  "image_build_engine": ""
}
```

**Reply by dan-jacobson** [reply]
WELP i deleted my `mint` and `mint-sensor` and downloaded the `1.41.7` release and now that exact command is working (shrug)

**Reply by kcq** [reply]
That's great that it's working now! There was an unfortunate combination of default flag values... The error says, `missing Docker connection info`, which means the app didn't handle the combination of flags involved in connecting to Docker. Most likely it's the `use tls` flag where it defaults to true.

**Reply by dan-jacobson** [reply]
Ok! I managed to get it working :). Final working command was `mint --host unix:///Users/drj/.orbstack/run/docker.sock slim --env-file .env --http-probe-cmd-file .mint/probeCmds.json my_image`.

One thing I noticed -- it seems that `--env-file` was not working the way I expected right now. my image makes a call to a postgres db, so I have an env var called `DB_PW` and another called `db_connection` like `postgresql://blah.blah:${DB_PW}@my-db.com/4321/postgres`. this works fine when i do `docker run --env-file .env my_image` but for some reason fails in `mint`. When I manually do the variable substitution it works no problem

**Comment by kcq** [comment]
The next version will have support for Docker contexts. In addition to defaulting to the current context it'll be possible to select a specific context like this: `mint --crt-context orbstack --quiet images` (it's there in master if you want to build the latest version).
